Rtl931x kernel 6.12 support

sharadanand opened the pull request.

Commit:
rtl93xx: i2c: Fix i2c multi byte read and add support for BLOCK transfers

Fixing multi byte read and add support for BLOCK transfer which was
missing.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-authored by: Harshal Gohel <hg@simonwunderlich.de>

Commit:
RTL931X: Fixing initialization of RTL9312 SoC.

Adding support for RTL9312 SoC.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
RTL931X: Fixing SoC probe for RTL931X SoCs.

While probing the SoC interrupt mask was wrong.
Fixing the serdes interrupt mask.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-authored-by: Harshal Gohel <hg@simonwunderlich.de>

Commit:
RTL931X: Additional support for RTL931X Soc.

- Adding crash protection by adding function pointer check.
- Fixing the mac config function for RTL931X.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
RTL931X: Fixing sysupgrade procedure.

Fixing kernel configuration to update /proc/cmdline parameter.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
rtl8224: Add initial probing support for rtl8224

Disable low power mode on phy to start, this removes requirement of "rtk
network on" before boot, however "rtk init" is still required

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-athored-by: Harshal Gohel <hg@simonwunderlich.de>
Co-athored-by: Simon Wunderlich <sw@simonwunderlich.de>

Commit:
rtl9310: read link status from rtl9310 serdes correctly

This forces ethtool to read correct link status on SFP cages

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-authored-by: Harshal Gohel <hg@simonwunderlich.de>

Commit:
rtl931x: Fixing VLAN tagging and untagging.

Correct the VLAN tagging and un-tagging table settings.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
rtl931x: Fixing serdes initialization.

Seting serdes registers correctly.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
rtl931x: Fixing chip init function.

MAC L2 layer register init fix.
Initialize the SPI part correctly.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
rtl931x: Improve traffic_get/set/enable/disable functions

- traffic_enable/disable assumed reverse byte order

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-authored-by: Harshal Gohel <hg@simonwunderlich.de>

hailfinger commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+// #define IN_BAND_STATUS
+
+/ {
+	compatible = "plasmacloud,psx28", "plasmacloud,esx28", "rapidmax,psx24", "realtek,rtl838x-soc";
```
Please mention the rapidmax psx24 in the changelog or drop it.

Djfe commented:
Thanks for adjusting your commits to the new testing kernel. The PR title is a bit misleading though. You are improving support for RTL931x on Kernel 6.12. (not adding it, like the title might imply)

__________________
#### Just at a quick glance: Some of the questions from the previous PR remain
- About commit https://github.com/openwrt/openwrt/pull/19173/commits/2a4297aa5265bbc1d349972cdec6a5a7f21b17c6: According to @plappermaul the old implementation was dropped intentionally and replaced with a new one. Please evaluate whether the replacement solves all your needs.

- About commit https://github.com/openwrt/openwrt/pull/19173/commits/798bcfabd01b23d61fa228c1e0c4ed00cec80160: Please help @jonasjelonek in verifying his PR https://github.com/openwrt/openwrt/pull/19171, approve it and then feel free to drop your version :)

- Parts of your commit https://github.com/openwrt/openwrt/pull/19173/commits/ab44e7eb3899acf93e238576c70baae1f3ee76cc might be merged first via https://github.com/openwrt/openwrt/pull/19170
Not sure who of you was first and honestly I shouldn't care. I would assume you both noticed the issue and found a solution independently.

__________________
Small nitpicks about your commit messages: OpenWrt prefers/wants commit messages and subjects to be written in imperative form (just like you did with "Improve")
adding > add
copying > copy
https://openwrt.org/submitting-patches#submission_guidelines
Also feel free to add a message to commits that outlines why a change is useful/necessary.

For adding device support, maintainers expect a commit message to contain extra info like in the following commit: https://github.com/openwrt/openwrt/commit/6af9476b8a3c75ce65f17cfa89da7d35360a4756
- specs
- installation instructions
- recovery instructions (optional)

It doesn't have to be as extensive as in the commit message above but every bit helps :)
How this looked for another plasma cloud device (though an AP not a switch): https://github.com/openwrt/openwrt/commit/4871fd2616acb03fefe69b068955dba36eb00770
Why is this helpful? Because such information will eventually vanish from OEM websites once a product goes EoL. But it will remain intact in OpenWrt commits and the OpenWrt wiki.

__________________
#### TL;DR
Everything you can split from this PR into smaller subsets **improves** the review process for everyone and keeps things **simple**.
- As an example I suggest creating an independent PR for the _Plasma Cloud - psx28/esx28_. It has nothing to do with improving support for the plattform overall (it might just depend on this PR to be merged first).
Adding a new device is far simpler and can be reviewed by committers that are not too familiar with the Realtek target.

I know this process is more tedious and requires more work on your part but it helps everyone involved with testing to help get your changes approved faster.
I'm sorry, if my long message overwhelms you. But it's the same for us with such a long PR:
Discussion goes on about different commits and people get confused what others are talking about. Splitting PR's help keep things straight and discussions short :)

If English or writing commit messages is not your strong point we could split that work. Lot's of people working on the target at the moment also speak German. So maybe you can message them somehow outside of this PR to collaborate on the target.
This would also prevent you and @jonasjelonek from sending in patches for similar/the same changes. You could even send them in together.

Neustradamus commented:
@sharadanand: Thanks for this PR which add this new support!

Djfe reviewed (commented):
To move this along, here's my review with things, I noticed.
These are small changes/suggestions, since I'm not familiar at all with most kernel code.

I missed in my earlier comment, that this is the first RTL931x device being added to OpenWrt, so the PR title is probably fine. (unlike what I wrote)
Thanks a lot for adding the first device! <3

Several commits actually have a commit message already, great. I only looked at a handful yesterday and made false assumptions, sorry.

This commit message appears to be missing a new line before "Fix"
https://github.com/openwrt/openwrt/pull/19173/commits/4c54c45faee6e00efda0fa0f8712e31cd6d84984

Looking at the specs: https://www.plasma-cloud.com/product-detail/psx28/
Since this device comes with a PoE Budget of 400W: It might be useful to configure the PoE budget in userspace when using [poemgr](https://github.com/blocktrron/poemgr) or add a new config file within the package to do so:
https://github.com/blocktrron/poemgr/blob/87550e1349c3c010a7ed0e306589c6398238d374/contrib/usw-lite.config#L4

Other realtek devices are using a different package called [realtek-poe](https://github.com/Hurricos/realtek-poe). For those it is configured here:
https://github.com/openwrt/openwrt/blob/9b777547beefa491319498208310f0758b49c45d/target/linux/realtek/base-files/etc/board.d/02_network#L111

Apparently your device comes with 3 fans, is it possible to set them up in the dts (and later configure them from userspace) or are they unconfigurable/connected to power directly?

0xharshal commented:
> https://github.com/openwrt/openwrt/blob/9b777547beefa491319498208310f0758b49c45d/target/linux/realtek/base-files/etc/board.d/02_network#L111
> 
> Apparently your device comes with 3 fans, is it possible to set them up in the dts (and later configure them from userspace) or are they unconfigurable/connected to power directly?

Yes devices uses adt7476 and already recognized by lm-sensors,
these fans are regulated using DC-PWM control, controlled on `pwm2` lane of the adt7476. 
We use configuration inspired by [target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol](https://github.com/openwrt/openwrt/blob/main/target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol), `seagate,blackarmor-nas220`

aparcar commented:
build realtek/rtl931x/plasmacloud_psx28

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+			};
+		#endif
+		};
+		port@52 {
+			reg = <52>;
+			label = "lan27";
+			phy-handle = <&phy26>;
+			sfp = <&sfp2>;
+			led-set = <1>;
+			phy-mode = "10gbase-r";
+		#if defined(IN_BAND_STATUS)
+			managed = "in-band-status";
+			// managed = "auto";
+		#else
+			fixed-link {
+				speed = <10000>;
+				full-duplex;
+				pause;
+			};
+		#endif
+		};
+		port@53 {
+			reg = <53>;
+			label = "lan28";
+			phy-handle = <&phy27>;
+			sfp = <&sfp3>;
+			led-set = <1>;
+			phy-mode = "10gbase-r";
+		#if defined(IN_BAND_STATUS)
+			managed = "in-band-status";
+			// managed = "auto";
+		#else
+			fixed-link {
+				speed = <10000>;
+				full-duplex;
+				pause;
+			};
+		#endif
+		};
+	#endif
```
Why is the plasmacloud PSX28 the only device which doesn't use in-band-status for the SFP ports? All other devices seem to use

```suggestion
		port@48 {
			reg = <48>;
			label = "lan25";
			phy-mode = "1000base-x";
			phy-handle = <&phy24>;
			sfp = <&sfp0>;
			led-set = <1>;
			managed = "in-band-status";
		};

		port@50 {
			reg = <50>;
			label = "lan26";
			phy-mode = "1000base-x";
			phy-handle = <&phy25>;
			sfp = <&sfp1>;
			led-set = <1>;
			managed = "in-band-status";
		};

		port@52 {
			reg = <52>;
			label = "lan27";
			phy-mode = "1000base-x";
			phy-handle = <&phy26>;
			sfp = <&sfp2>;
			led-set = <1>;
			managed = "in-band-status";
		};

		port@53 {
			reg = <53>;
			label = "lan28";
			phy-mode = "1000base-x";
			phy-handle = <&phy27>;
			sfp = <&sfp3>;
			led-set = <1>;
			managed = "in-band-status";
		};
```

sharadanand replied:
Thank you for the feedback and updated in newer version

ecsv commented on target/linux/realtek/patches-6.12/805-Improve-SFP-support-for-10G-BX-and-LX-modules.patch:
```diff
@@ -0,0 +1,84 @@
++	pr_err("%s: %d fixing up 1000base_lx sfp\n", __func__, __LINE__);
++	sfp->id.base.e1000_base_t = false;
++	sfp->id.base.e1000_base_sx = false;
++	sfp->id.base.e1000_base_lx = true;
++	sfp->id.base.e1000_base_cx = false;
++
++	sfp->id.base.e10g_base_sr = false;
++	sfp->id.base.e10g_base_lr = false;
++	sfp->id.base.e10g_base_lrm = false;
++	sfp->id.base.e10g_base_er = false;
++
++}
++
++static void sfp_fixup_10gbase_lr(struct sfp *sfp)
++{
++	pr_err("%s: %d fixing up 10gbase-lr sfp\n", __func__, __LINE__);
++	sfp->id.base.e1000_base_t = false;
++	sfp->id.base.e1000_base_sx = false;
++	sfp->id.base.e1000_base_lx = false;
++	sfp->id.base.e1000_base_cx = false;
++
++	sfp->id.base.e10g_base_sr = false;
++	sfp->id.base.e10g_base_lr = true;
++	sfp->id.base.e10g_base_lrm = false;
++	sfp->id.base.e10g_base_er = false;
++}
++
++
+ #define SFP_QUIRK(_v, _p, _m, _f) \
+ 	{ .vendor = _v, .part = _p, .modes = _m, .fixup = _f, }
+ #define SFP_QUIRK_M(_v, _p, _m) SFP_QUIRK(_v, _p, _m, NULL)
+@@ -523,6 +572,10 @@ static const struct sfp_quirk sfp_quirks[] = {
+ 	SFP_QUIRK_F("Turris", "RTSFP-2.5G", sfp_fixup_rollball),
+ 	SFP_QUIRK_F("Turris", "RTSFP-10", sfp_fixup_rollball),
+ 	SFP_QUIRK_F("Turris", "RTSFP-10G", sfp_fixup_rollball),
++	SFP_QUIRK_F("Mikrotik", "S-RJ01", sfp_fixup_1000base_lx),
++	SFP_QUIRK_F("CISCO-PROLABS", "WAVS+BX32-10CCN2", sfp_fixup_10gbase_lr),
++	SFP_QUIRK_F("CISCO-PROLABS", "WAVS+BX23-10CCN2", sfp_fixup_10gbase_lr),
++	SFP_QUIRK("DUMMY", "DUMMY", sfp_quirk_10gbase_lr, sfp_fixup_10gbase_lr),
+ };
```
Wow, the suggestion diff is quite bad. I basically removed `sfp_quirk_10gbase_lr` and `sfp_quirk_10gbase_lr`.

ecsv replied:
The dummy code is still there.

ecsv commented on target/linux/realtek/files-6.12/drivers/gpio/gpio-rtl8231.c:
```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/gpio/driver.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/delay.h>
+#include <asm/mach-rtl838x/mach-rtl83xx.h>
+
+/* RTL8231 registers for LED control */
+#define RTL8231_LED_FUNC0			0x0000
+#define RTL8231_LED_FUNC1			0x0001
+#define RTL8231_READY_MASK			0x03f0
+#define RTL8231_READY_VALUE			0x0370
+#define RTL8231_GPIO_PIN_SEL(gpio)		((0x0002) + ((gpio) >> 4))
+#define RTL8231_GPIO_DIR(gpio)			((0x0005) + ((gpio) >> 4))
+#define RTL8231_GPIO_DATA(gpio)			((0x001C) + ((gpio) >> 4))
```
This driver was readded after it was removed in b410f2216c7a4e6c5646f999d18b952b63c85817. Maybe it should be converted instead. See f31c9bb237c306c89b553a96c96460f194202df2 for details about the conversion

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+		};
+	};
+
+	gpio1: rtl8231-gpio {
+		compatible = "realtek,rtl8231-gpio";
+		#gpio-cells = <2>;
+		gpio-controller;
+		indirect-access-bus-id = <24>;
+	};
```
This is outdated and needs to be converted. Run

```
git log -p -Srealtek,rtl8231-gpio origin/main -- target/linux/realtek/dts/
```

to see a couple of examples

jonasjelonek commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+	led_set: led_set@0 {
+		compatible = "realtek,rtl9300-leds";
+		active-low;
+		num_led_per_port = <2>;
```
No, this `leds_on_this_port` is currently derived from the array size of `led_setX`, see https://github.com/openwrt/openwrt/blob/8ad5416d997312a4ae86ebb7bee7dfd1f80c0407/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/common.c#L411

However, this is currently non-functional for rtl931x because the driver part doesn't use this, see https://github.com/openwrt/openwrt/blob/8ad5416d997312a4ae86ebb7bee7dfd1f80c0407/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c#L1580-L1581

I already have patch in my local tree for this, just needs to be submitted :)

jonasjelonek replied:
see #19241

ecsv replied:
This problem is still there

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+
+		/* fiber */
+	#if defined(ENABLE_SFP_PHYS)
+		phy24: ethernet-phy@48 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			phy-is-integrated;
+			reg = <48>;
+			rtl9300,smi-address = <0 0>;
+			sds = <8>;
+		};
+
+		phy25: ethernet-phy@50 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			phy-is-integrated;
+			reg = <50>;
+			rtl9300,smi-addresss = <0 0>;
+			sds = <9>;
+		};
+
+		phy26: ethernet-phy@52 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			phy-is-integrated;
+			reg = <52>;
+			rtl9300,smi-address = <0 0>;
+			sds = <10>;
+		};
+
+		phy27: ethernet-phy@53 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			phy-is-integrated;
+			reg = <53>;
+			rtl9300,smi-address = <0 0>;
+			sds = <11>;
+		};
```
Please check if you really need to use clause 45 for the integrated phy of the SerDes.

Usually, I would expected to see following block:


```suggestion
		INTERNAL_PHY_SDS(48, 8)
		INTERNAL_PHY_SDS(50, 9)
		INTERNAL_PHY_SDS(52, 10)
		INTERNAL_PHY_SDS(53, 11)
```

But this would then configure it as `ethernet-phy-ieee802.3-c22` - which all other devices use for the "same" SerDes.

0xharshal commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+	model = "Plasma Cloud PSX28";
+
+	chosen {
+		bootargs = "earlycon loglevel=8 nowait";
```
The device uses bootargs to make dual partition boot work, for that we use 
`CONFIG_MIPS_CMDLINE_DTB_EXTEND=y` so bootargs are extended and not replaced.

I still agree with you, default bootargs are not serving any purpose yet.

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+
+		};
+
+		i2c06: i2c-poe {
+			reg = <6>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			sda-pin = <20>;
+			scl-pin = <13>;
+		};
```
The `plasmacloud,esx28` doesn't have any POE capabilities. Maybe split the DTS up in a common `*.dtsi`. Both `rtl9312_plasmacloud_psx28.dts` and `rtl9312_plasmacloud_esx28.dts` should include the common `rtl9312_plasmacloud_xsx28.dtsi` and both at least define their own `compatible` and `model`. Something like:

```
/ {
	compatible = "plasmacloud,psx28", "realtek,rtl931x-soc";
	model = "Plasma Cloud PSX28";
};
```

(please correct me about the least significant compatible name - it is not really handled consistently in the `rtl930*.dts` files)

And nodes specific to `plasmacloud,psx28` must then placed in has to the `rtl9312_plasmacloud_psx28.dts`. This can look like this:

```
/ {
	compatible = "plasmacloud,psx28", "realtek,rtl931x-soc";
	model = "Plasma Cloud PSX28";

	i2cmux {
		i2c06: i2c-poe {
			reg = <6>;
			#address-cells = <1>;
			#size-cells = <0>;
			sda-pin = <20>;
			scl-pin = <13>;
		};
	};

	leds {
		led-2 {
			color = <LED_COLOR_ID_GREEN>;
			function = "alarm-poe";
			gpios = <&gpio1 28 GPIO_ACTIVE_LOW>;
		};
	};
};
```

Or if you create a handle to the mux in the `dtsi`:

```
/ {
	compatible = "plasmacloud,psx28", "realtek,rtl931x-soc";
	model = "Plasma Cloud PSX28";

       ....
};

&i2cmux0 {
	i2c-poe {
		reg = <6>;
		#address-cells = <1>;
		#size-cells = <0>;
		sda-pin = <20>;
		scl-pin = <13>;
	};
};
```

Commit:
realtek: add support for plasma cloud - psx28/esx28

Device specification -

SoC - Realtek RTL9312

CPU Model - MIPS interAptiv (multi) V2.0.

Memory - 256 MB.

Flash memory - 32 MB NOR.

Installation procedure -

Via U-boot:

 1. Switch ON the hardware using - rtk network on
 2. Set server IP address - setenv serverip 192.168.1.1
 3. Set IP address - setenv ipaddr 192.168.1.20
 4. Get the RAMfs image into image - tftpboot 0x84000000 openwrt-realtek-rtl931x-plasmacloud_psx28-initramfs-kernel.bin
 5. Boot from the RAM - bootm 0x84000000

Image will be booted and can be flashed by using sysupgrade.

Via sysupgrade:
 1. Transfer the sysupgrade image to the switch -
 	scp openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin root@[IP V6 address]:/tmp/

 2. Perform system update using sysupgrade command -
 	sysupgrade -n /tmp/openwrt-realtek-rtl931x-plasmacloud_psx28-squashfs-sysupgrade.bin

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
RTL931x: Add GPIO access register definition for RTL931X SoC.

Define GPIO acceess register definition.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
RTL931X: Improve SFP support for 10G-BX and LX modules.

Patching kernel to improve SFP modules.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

Commit:
RTL931X: Disable callbacks for l3 hw routing for rtl9310.

Add experimental support for eee Fix a bug where rtl9310 serdes is probed as rtl9300

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Co-authored-by: Harshal Gohel <hg@simonwunderlich.de>

Commit:
rtl931x: Fixing missing function pointers.

Fixing smi_poll_ctrl and l2_hash_seed function pointer initialization.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>

jonasjelonek commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c:
```diff
@@ -1688,6 +1692,7 @@ const struct rtl838x_reg rtl931x_reg = {
 	.pie_rule_write = rtl931x_pie_rule_write,
 	.pie_rule_add = rtl931x_pie_rule_add,
 	.pie_rule_rm = rtl931x_pie_rule_rm,
+	.l2_hash_seed = rtl931x_l2_hash_seed,
```
This is now duplicated in the code because of 24d9fdff2e. Please drop this and the `rtl931x_l2_hash_seed` too (mentioned commit renamed it to `rtldsa_931x_l2_hash_seed`)

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dts:
```diff
@@ -0,0 +1,700 @@
+
+	chosen {
+		bootargs = "earlycon loglevel=8 nowait";
+	};
```
still missing

ecsv commented on target/linux/realtek/dts/rtl931x.dtsi:
```diff
@@ -205,6 +214,10 @@
 			pinctrl-single,bits = <0x0 0x0 0x100>;
 		};
+
+		pinmux_gpio_mdio_en: gpio-mdio-en {
```
why is there now this change which adds `mdio-aux` and `gpio-mdio-en` for all rtl931x devices in a commit which is supposed to add only PSX28/ESX28 support?

And it is also using a compatibility string which is for a completely different SoC. Please clean this up and make it a separate commit for adding `realtek_aux_mdio_driver` to the driver and the shared `rtl931x.dtsi`. See d45890f6cd508256356f99c11b8a744a30d99cf4 for a commit which adds it tot he rtl930x.dtsi

ecsv commented on target/linux/realtek/dts/rtl9312_plasmacloud_psx28.dtsi:
```diff
@@ -0,0 +1,676 @@
+	};
+};
+
+&ethernet0 {
```
Thanks for the info. In that case, @sharadanand must make sure that the mac addresses for ports (and the main mac) is set correctly in `target/linux/realtek/base-files/etc/board.d/02_network`

Neustradamus commented:
@sharadanand: Can you publish here all new PRs?

ecsv commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -3917,6 +3925,21 @@ static struct phy_driver rtl83xx_phy_driver[] = {
+		PHY_ID_MATCH_MODEL(PHY_ID_RTL8224),
+		.name		= "REALTEK RTL8224",
+		.features	= PHY_10GBIT_FEATURES,
+		.flags		= PHY_POLL_CABLE_TEST,
```
There is no cable test functionality. And I would have expected the usage of `rtl821x_read_page` and `rtl821x_write_page`

Djfe commented:
quite cool to see 10/15 PRs already merged in 17 days since then.
While one has just been superseeded and one was put on hold to work on redoing how mdio works.
Only 3 or 4 PRs of the original submission are left ☺️

Djfe commented:
Looking forward to the next PR for the PSX28/ESX28 💪
Thanks for all your (plural) work on the realtek target

ecsv commented:
> > The PSX28+ESX28 support is currently still missing. Some of the userspace helpers can be found in the PSX10 support PR #19362. I would therefore consider the PSX10 support as dependency for the PSX28/ESX28 support.
> 
> @ecsv do you plan to submit a PR for the Plasma Cloud PSX28 now that #19362 is merged?

I finally found some time for this. The basic support can be found in #20172
